### ISSUE TITLE
projectile-switch-project does not immediately changes the current buffer

### ISSUE THREAD

**Issue by xl666** [bbatsov/projectile#1617]
I think I found a bug in the latest version of Projectile.
projectile-switch-project does not immediately changes the current
buffer to the one selected in the switching, this is a problem for me
because I do the switching as part of some custom code related to my
workflow. This problem appeared after the last update.

For example, if I'm on the *scratch* buffer and I execute:

(progn (projectile-switch-project) (current-buffer))

The return value is #<buffer *scratch*> instead of whatever buffer is
opened as part of project switching

Doing some hacking, I solved the problem in my code as follows:

(progn (projectile-switch-project)
  (switch-to-buffer (car (buffer-list)))
  (current-buffer))

But is a hacky and ugly solution.

Thanks in advance.

Projectile version information:
20201204.954

Emacs version:

27.1

Operating system:

Linux kernel 5.9

**Comment by bbatsov** [comment]
If I had to guess, I'd say it's probably related to this change https://github.com/bbatsov/projectile/commit/5406793bee52f7005f20a78cf01b064dcbb6dbbc I'll take a closer look at the problem when I can. In the mean time maybe @kevinjfoley or @Deewiant will beat me to it.

**Comment by kevinjfoley** [comment]
@bbatsov - you're right it is related to that change.  Something about changing buffers in a temp buffer doesn't carry over in the same form.  Here's a simplified example that demonstrates the issue:
```
;; call from *scratch*
(progn
  (with-temp-buffer
    (switch-to-buffer "*Messages*")
    ;; First buffer *Messages*
    (message "First buffer %s" (buffer-name)))
  ;; Second buffer *scratch*
  (message "Second buffer %s" (buffer-name)))
```

Someone who understands the mechanics of switching buffers better and how `with-temp-buffer` impacts that may know a better fix but here's my hacky solution:
```
(progn
  (let (buffer)
    (with-temp-buffer
      (switch-to-buffer "*Messages*")
      (message "First buffer %s" (buffer-name))
      (setq buffer (current-buffer)))
    (with-current-buffer buffer
      (message "Second buffer %s" (buffer-name)))))
```

**Comment by Deewiant** [comment]
`with-temp-buffer` is implemented in terms of `save-current-buffer`:

> Record which buffer is current; execute BODY; make that buffer current.

So it makes sense that buffer switches done within the body are lost. @kevinjfoley's solution is what I'd've suggested as well.

Though at this point I'm confused as to why 5406793bee52f700 works at all, since it seems like `switch-project-action`'s buffer switch should be completely overridden by `with-temp-buffer`.

**Comment by verdammelt** [comment]
I'm seeing an issue (not sure when it started, but quite recent) where `default-directory` is not the project root as I'd expect in a function I have added to `projectile-after-switch-project-hook`. I'm guessing that would be the same as this issue (if the buffer has not updated then the default directory will not). 

If it seems like what I'm seeing is a different thing. I'll create another issue for it.

**Comment by bbatsov** [comment]
99% your issue is caused by the same change. I'm starting to wonder if we shouldn't revert it for now.

**Comment by Deewiant** [comment]
@bbatsov Any reason why the solution of taking `(current-buffer)` explicitly as shown above wouldn't work?

5406793bee is a big usability improvement for me so I'd be sad to see it reverted so soon.

**Comment by bbatsov** [comment]
@Deewiant The proposed solution seems a bit dicey, but I'll willing to give it a go. PR welcome!

**Comment by kevinjfoley** [comment]
> Any reason why the solution of taking (current-buffer) explicitly as shown above wouldn't work?

I realized we could also just run the hooks from within the `with-temp-buffer` form as well.  I think that should handle @verdammelt 's issue.

However I'm not sure it will handle @xl666 's example where a function is called in the same form as `projectile-switch-project`.  It seems the changes made in the `with-temp-buffer` won't apply until after the form exits.  I'm not sure what the solution for that is.

**Comment by Deewiant** [comment]
I opened PR #1621 using the strategy of explicitly switching the buffer. It should address all three use cases (i.e. the two shown here plus it doesn't break 5406793).

**Comment by bbatsov** [comment]
Fixed by #1621.
